onvif/src/discovery/mod.rs

```diff
@@ -83,79 +111,110 @@ impl DiscoveryBuilder {
-    ///
-    /// async {
-    ///     discovery::DiscoveryBuilder::default().run()
-    ///         .await
-    ///         .unwrap()
-    ///         .for_each_concurrent(MAX_CONCURRENT_JUMPERS, |addr| {
-    ///             async move {
-    ///                 println!("Device found: {:?}", addr);
-    ///             }
-    ///         })
-    ///         .await;
-    /// };
-    /// ```
-    ///
-    /// Or you can await on a collection of unique devices found in one second:
-    ///
-    /// ```
-    /// use onvif::discovery;
-    /// use futures_util::stream::StreamExt; // to use collect
-    /// use std::collections::HashSet;
-    ///
-    /// async {
-    ///     let devices = discovery::DiscoveryBuilder::default().run()
-    ///         .await
-    ///         .unwrap()
-    ///         .collect::<HashSet<_>>()
-    ///         .await;
-    ///
-    ///     println!("Devices found: {:?}", devices);
-    /// };
-    /// ```
-    pub async fn run(&self) -> Result<impl Stream<Item = Device>, Error> {
-        let Self {
-            duration,
-            listen_address,
-        } = self;
+    /// Set the discovery mode. See [DiscoveryMode] for a description of how this works.
+    /// By default, the multicast mode is chosen.
+    pub fn discovery_mode(&mut self, discovery_mode: DiscoveryMode) -> &mut Self {
+        self.discovery_mode = discovery_mode;
+        self
+    }
 
+    async fn run_unicast(
+        &self,
+        duration: &Duration,
+        listen_address: &IpAddr,
+        network: &Ipv4Addr,
+        network_mask: &Ipv4Addr,
+    ) -> Result<ReceiverStream<Device>, Error> {
         let probe = Arc::new(build_probe());
         let probe_xml = yaserde::ser::to_string(probe.as_ref()).map_err(Error::Serde)?;
 
-        debug!("Probe XML: {}", probe_xml);
+        debug!("Unicast Probe XML: {}. Since you are using unicast, some devices might not be detected", probe_xml);
 
-        let socket = {
-            const LOCAL_PORT: u16 = 0;
+        let message_id = Arc::new(probe.header.message_id.clone());
+        let payload = Arc::new(probe_xml.as_bytes().to_vec());
+
+        let (device_sender, device_receiver) = channel(32);
+        let device_receiver = ReceiverStream::new(device_receiver);
+
+        let mut unicast_requests = vec![];
+
+        for target_address in enumerate_network_v4(*network, *network_mask) {
+            let local_socket_addr = SocketAddr::new(*listen_address, Self::LOCAL_PORT);
+            let target_sock_addr = SocketAddr::new(IpAddr::V4(target_address), Self::MULTI_PORT);
+
+            let socket = UdpSocket::bind(local_socket_addr).await?;
+
+            unicast_requests.push((
+                socket,
+                target_sock_addr,
+                device_sender.clone(),
+                payload.clone(),
+                message_id.clone(),
+            ));
+        }
+
+        let produce_devices = async move {
+            futures::future::join_all(unicast_requests.iter().map(
```

Comment:
Let's limit the concurrency when joining futures. Otherwise, we may run out of resources if too many IPs were enumerated.

https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffer_unordered

Reply:
I don't think that would work with the lookup limit since some IPs will never reply. I'll try to adapt the code in a way that handles both the volume and the duration timeout.

Will keep you posted at it might not be today

Reply:
I don't think that would work with the lookup limit since some IPs will never reply. I'll try to adapt the code in a way that handles both the volume and the duration timeout.

Will keep you posted at it might not be today